pkg/virtctl/credentials/add-key/add-key_test.go

```diff
@@ -0,0 +1,562 @@
+
+		secret.OwnerReferences = nil
+
+		_, err = kubeClient.CoreV1().Secrets(metav1.NamespaceDefault).Update(context.Background(), secret, metav1.UpdateOptions{})
+		Expect(err).ToNot(HaveOccurred())
+
+		err = runAddKeyCommand(
+			"--user", userName,
+			"--value", "test-key",
+			vmName,
+		)
+		Expect(err).To(MatchError(ContainSubstring("does not have an owner reference pointing to VM")))
+	})
+
+	It("should patch secret not owned by VM, with --force option", func() {
+		secret, err := kubeClient.CoreV1().Secrets(metav1.NamespaceDefault).Get(context.Background(), secretName, metav1.GetOptions{})
+		Expect(err).ToNot(HaveOccurred())
+
+		secret.OwnerReferences = nil
+
+		_, err = kubeClient.CoreV1().Secrets(metav1.NamespaceDefault).Update(context.Background(), secret, metav1.UpdateOptions{})
+		Expect(err).ToNot(HaveOccurred())
+
+		const testKey = "test-key"
+		err = runAddKeyCommand(
+			"--user", userName,
+			"--value", testKey,
+			"--force",
+			vmName,
+		)
+		Expect(err).ToNot(HaveOccurred())
+
+		secret, err = kubeClient.CoreV1().Secrets(metav1.NamespaceDefault).Get(context.Background(), secretName, metav1.GetOptions{})
+		Expect(err).ToNot(HaveOccurred())
+
+		Expect(secret.Data).To(HaveLen(1))
+		Expect(secret.Data).To(ContainElement([]byte(testKey)))
+	})
+
+	Context("with --create-secret set", func() {
+		It("should create a new secret and patch VM", func() {
+			// VM should not be running
+			vmi = nil
+
+			const testKey = "test-key"
+			err := runAddKeyCommand(
+				"--user", userName,
+				"--value", testKey,
+				"--create-secret",
+				vmName,
+			)
+			Expect(err).ToNot(HaveOccurred())
+
+			Expect(vm.Spec.Template.Spec.AccessCredentials).To(HaveLen(2))
+			credential := vm.Spec.Template.Spec.AccessCredentials[1]
+			Expect(credential.SSHPublicKey).ToNot(BeNil())
+			Expect(credential.SSHPublicKey.Source.Secret).ToNot(BeNil())
+			Expect(credential.SSHPublicKey.Source.Secret.SecretName).ToNot(BeEmpty())
+			Expect(credential.SSHPublicKey.PropagationMethod.QemuGuestAgent).ToNot(BeNil())
+			Expect(credential.SSHPublicKey.PropagationMethod.QemuGuestAgent.Users).To(ContainElement(userName))
+
+			newSecretName := credential.SSHPublicKey.Source.Secret.SecretName
+
+			secret, err := kubeClient.CoreV1().Secrets(metav1.NamespaceDefault).Get(context.Background(), newSecretName, metav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			Expect(secret.Data).To(HaveLen(1))
+			Expect(secret.Data).To(ContainElement([]byte(testKey)))
+		})
+
+		It("should patch VM with nil AccessCredentials slice", func() {
+			// VM should not be running
+			vmi = nil
+
+			vm.Spec.Template.Spec.AccessCredentials = nil
+
+			const testKey = "test-key"
+			err := runAddKeyCommand(
+				"--user", userName,
+				"--value", testKey,
+				"--create-secret",
+				vmName,
+			)
+			Expect(err).ToNot(HaveOccurred())
+
+			Expect(vm.Spec.Template.Spec.AccessCredentials).To(HaveLen(1))
+			credential := vm.Spec.Template.Spec.AccessCredentials[0]
```

Comment:
same comment as above, try to collect the statements that repeat into a function. It improves the readability and avoid repetitions

Reply:
done